### PR TITLE
Added PNGPanel as a base class and for _repr_png_

### PR DESCRIPTION
**Pull request by jbednar**
Previously, MatplotlibPanel and RGGPlotPanel had separate implementations for rendering PNGs into something Viewable.  This PR pulls that functionality into a new shared base class PNGPanel that makes those classes much simpler (Matplotlib support is now essentially just one call).  

PNGPanel isn't an abstract class, though; it directly renders anything with a `_repr_png_` method, which means that it supports PIL/Pillow objects, Datashader objects, and probably quite a lot of other objects that are designed to display themselves as PNGs in Jupyter.   

PNGPanel computes its size directly from the PNG, which means that it no longer needs an MPL-specific call and Datashader plots no longer need to specify their size (as it's in the PNG already).  So we can now do:

```
import numpy as np, pandas as pd, datashader as ds
from datashader import transfer_functions as tf
from colorcet import kbc
from numba import jit

@jit
def clifford_trajectory(a, b, c, d, x0, y0, n):
    xs, ys = np.zeros(n), np.zeros(n)
    xs[0], ys[0] = x0, y0
    for i in np.arange(n-1):
        xs[i+1] = np.sin(a * ys[i]) + c * np.cos(a * xs[i])
        ys[i+1] = np.sin(b * xs[i]) + d * np.cos(b * ys[i])
    return pd.DataFrame(dict(x=xs,y=ys))

def clifford_plot(a, b, c, d, n=10000000, cmap=kbc[::-1]):
    cvs = ds.Canvas(plot_width=400, plot_height=400)
    agg = cvs.points(clifford_trajectory(a, b, c, d, 0, 0, n), 'x', 'y')
    return tf.shade(agg, cmap=cmap)
```

```
from PIL import Image
logo = Image.open("../../datashader/doc/_static/logo_stacked.png").resize((200,150))
```

```
from panel import Row, Column
import param

class DatashaderPanel(param.Parameterized):
    a = param.Number(1.7, bounds=(-2, 2))
    b = param.Number(1.7, bounds=(-2, 2))
    c = param.Number(0.6, bounds=(-2, 2))
    d = param.Number(1.2, bounds=(-2, 2))

    @param.depends('a', 'b', 'c', 'd')
    def view(self):
        return clifford_plot(self.a, self.b, self.c, self.d)
panel = DatashaderPanel()
Row(Column(logo, panel), panel.view)
```

image

### CODE REVIEW

File: panel/panels.py [hunk 1]
```diff
@@ -327,32 +327,57 @@ def _update(self, model):
...
         b64 = base64.b64encode(data).decode("utf-8")
         src = "data:image/png;base64,{b64}".format(b64=b64)
-        width, height = self.object.canvas.get_width_height()
         html = "<img src='{src}'></img>".format(src=src)
         
-        p = super(MatplotlibPanel,self)._get_properties()
+        p = super(PNGPanel,self)._get_properties()
+        width, height = self._pngshape(data)
         if self.width  is None: p["width"]  = width
         if self.height is None: p["height"] = height
```
Comment (philippjfr) [inline panel/panels.py]: This should be:

```python
p['width'] = self.width or width
p['height'] = self.height or height
```

The width and height should always be provided.

**Comment by philippjfr** [comment]
Looks great, the ``struct`` trick to get the png width and height is really neat.

File: panel/panels.py (hunk 1 above)
Comment (philippjfr) [reply panel/panels.py]: Nevermind it's fine I see the superclass defines it.

Comment (jbednar) [reply panel/panels.py]: Shouldn't it be possible for the user to set self.width or self.height to zero, if they want to?  The proposed replacement code doesn't allow that.

Comment (jbednar) [reply panel/panels.py]: (Since self.width and self.height are only for user overriding of the computed width,height).

Comment (jbednar) [reply panel/panels.py]: Oh, I see; you mean that `p["width"]` and `p["height"]` should always be provided.  So I guess it should be:

```
p["width"]  = width  if self.width  is None else self.width
p["height"] = height if self.height is None else self.height
```

Comment (jbednar) [reply panel/panels.py]: Oops; I never saw your "nevermind"; somehow the PR page updated with only your first comment.

**Pull request merged**